src/nlp/search.py

```diff
@@ -0,0 +1,435 @@
+if TYPE_CHECKING:
+    from .arrow_dataset import Dataset  # noqa: F401
+
+
+try:
+    import elasticsearch as es
+    from elasticsearch import Elasticsearch
+    import elasticsearch.helpers
+
+    _has_elasticsearch = True
+except ImportError:
+    _has_elasticsearch = False
+
+try:
+    import faiss
+
+    _has_faiss = True
+except ImportError:
+    _has_faiss = False
+
+
+logger = logging.getLogger(__name__)
+
+
+class MissingIndex(Exception):
+    pass
+
+
+SearchResults = NamedTuple("SearchResults", [("scores", List[float]), ("indices", List[int])])
+BatchedSearchResults = NamedTuple(
+    "SearchResults", [("total_scores", List[List[float]]), ("total_indices", List[List[int]])]
+)
+
+NearestExamplesResults = NamedTuple("SearchResults", [("scores", List[float]), ("examples", List[dict])])
+BatchedNearestExamplesResults = NamedTuple(
+    "SearchResults", [("total_scores", List[List[float]]), ("total_examples", List[List[dict]])]
+)
+
+
+class BaseIndex:
+    """Base class for indexing"""
+
+    def search(self, query, k: int = 10) -> SearchResults:
+        """
+        To implement.
+        This method has to return the scores and the indices of the retrieved examples given a certain query.
+        """
+        raise NotImplementedError
+
+    def search_batch(self, queries, k: int = 10) -> BatchedSearchResults:
+        """ Find the nearest examples indices to the query.
+
+            Args:
+                `queries` (`Union[List[str], np.ndarray]`): The queries as a list of strings if `column` is a text index or as a numpy array if `column` is a vector index.
+                `k` (`int`): The number of examples to retrieve per query.
+
+            Ouput:
+                `total_scores` (`List[List[float]`): The retrieval scores of the retrieved examples per query.
+                `total_indices` (`List[List[int]]`): The indices of the retrieved examples per query.
+        """
+        total_scores, total_indices = [], []
+        for query in queries:
+            scores, indices = self.search(query, k)
+            total_scores.append(scores)
+            total_indices.append(indices)
+        return BatchedSearchResults(total_scores, total_indices)
+
+
+class ElasticSearchIndex(BaseIndex):
+    """
+    Sparse index using Elasticsearch. It is used to index text and run queries based on BM25 similarity.
+    An Elasticsearch server needs to be accessible, and a python client is declared with
+    ```
+    es_client = Elasticsearch([{'host': 'localhost', 'port': '9200'}])
+    ```
+    for example.
+    """
+
+    def __init__(
+        self,
+        host: Optional[str] = None,
+        port: Optional[int] = None,
+        es_client: Optional[Elasticsearch] = None,
+        index_name: Optional[str] = None,
+        index_config: Optional[dict] = None,
+    ):
+        assert (
+            _has_elasticsearch
+        ), "You must install ElasticSearch to use ElasticSearchIndex. To do so you can run `pip install elasticsearch`"
```

Comment:
Maybe, to be more safe, we should propose a version of ES with which we know it works because I know by experience with ES that you can have big breaking changes across versions, even sometime across minor versions. And add it as well in the documentation.

This is really nice!! LGTM :)